internal/pki/encoding_test.go

```diff
@@ -15,340 +13,182 @@ package pki
+	t.Run("EncodedEmpty", func(t *testing.T) {
+		txt := []byte("-----BEGIN CERTIFICATE-----\n\n-----END CERTIFICATE-----\n")
 
-				if err := c.UnmarshalText(encoded); !errors.Is(err, ErrInvalidPEM) {
-					t.Fatalf("expected invalid PEM error")
-				}
-			})
-		})
+		var sink Certificate
+		assert.ErrorContains(t, sink.UnmarshalText(txt), "malformed")
 	})
-}
 
-func TestNewPrivateKey(t *testing.T) {
-	key, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-	privateKey := NewPrivateKey(key)
+	t.Run("EncodedGarbage", func(t *testing.T) {
+		txt := []byte("-----BEGIN CERTIFICATE-----\nasdfasdf\n-----END CERTIFICATE-----\n")
 
-	if reflect.TypeOf(privateKey).String() != "*pki.PrivateKey" {
-		t.Fatalf("expected *pki.PrivateKey in return")
-	}
-}
-
-func TestParseCertificate(t *testing.T) {
-	// generateCertificate is a helper function that generates a random private key
-	// and ignore any errors. creates a self-signed certificate as we don't need
-	// much
-	generateCertificate := func() *Certificate {
-		key, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-		now := time.Now()
-		template := &x509.Certificate{
-			BasicConstraintsValid: true,
-			KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
-			NotBefore:             now,
-			NotAfter:              now.Add(12 * time.Hour),
-			SerialNumber:          big.NewInt(1234),
-			SignatureAlgorithm:    certificateSignatureAlgorithm,
-			Subject: pkix.Name{
-				CommonName: "*",
-			},
-		}
-
-		certificate, _ := x509.CreateCertificate(rand.Reader, template, template, key.Public(), key)
-
-		return &Certificate{Certificate: certificate}
-	}
-
-	t.Run("valid", func(t *testing.T) {
-		expected := generateCertificate()
-		encoded := pem.EncodeToMemory(&pem.Block{Bytes: expected.Certificate, Type: pemCertificateType})
-
-		certificate, err := ParseCertificate(encoded)
-
-		if err != nil {
-			t.Fatalf("expected no error, actual %s", err.Error())
-		}
-
-		if !reflect.DeepEqual(expected.Certificate, certificate.Certificate) {
-			t.Fatalf("expected parsed certificate to match expected")
-		}
+		var sink Certificate
+		assert.ErrorContains(t, sink.UnmarshalText(txt), "malformed")
 	})
 
-	t.Run("invalid", func(t *testing.T) {
-		data := []byte("bad")
+	t.Run("ReadByOpenSSL", func(t *testing.T) {
+		openssl := require.OpenSSL(t)
+		dir := t.TempDir()
 
-		certificate, err := ParseCertificate(data)
+		certFile := filepath.Join(dir, "cert.pem")
+		certBytes, err := cert.MarshalText()
+		assert.NilError(t, err)
+		assert.NilError(t, os.WriteFile(certFile, certBytes, 0o600))
 
-		if err == nil {
-			t.Fatalf("expected error")
-		}
+		// The "openssl x509" command parses X.509 certificates.
+		cmd := exec.Command(openssl, "x509",
+			"-in", certFile, "-inform", "PEM", "-noout", "-text")
 
-		if certificate != nil {
-			t.Fatalf("expected certificate to be nil")
-		}
+		output, err := cmd.CombinedOutput()
+		assert.NilError(t, err, "%q\n%s", cmd.Args, output)
```

Comment:
Are we printing the cmd.Args here to make sure the `certFile` is what we expect it to be or for some other reason?

Reply:
To show the command that failed. The `openssl pkey` on my Mac, for example, doesn't support the `-check` argument. Removing the skip in that similar test fails like this:

```
--- FAIL: TestPrivateKeyTextMarshaling (0.01s)
    --- FAIL: TestPrivateKeyTextMarshaling/ReadByOpenSSL (0.01s)
        encoding_test.go:158: using "/usr/bin/openssl"
            LibreSSL 2.8.3
            built on: date not available
            platform: information not available
            options:  bn(64,64) rc4(16x,int) des(idx,cisc,16,int) blowfish(idx)
            compiler: information not available
            OPENSSLDIR: "/private/etc/ssl"
        --- FAIL: TestPrivateKeyTextMarshaling/ReadByOpenSSL/Check (0.01s)
            encoding_test.go:187: assertion failed: error is not nil: exit status 1: ["/usr/bin/openssl" "pkey" "-check" "-in" "/var/folders/bw/_dxdzsmx6h1ffcgp6rvfh55c0000gn/T/TestPrivateKeyTextMarshalingReadByOpenSSL1838345719/001/key.pem" "-inform" "PEM" "-noout" "-text"]
                Unknown cipher check
                Usage pkey [options]
                where options are
                -in file        input file
                -inform X       input format (DER or PEM)
                -passin arg     input file pass phrase source
                -outform X      output format (DER or PEM)
                -out file       output file
                -passout arg    output file pass phrase source
```